xrpl/transaction/main.py

```diff
@@ -79,7 +105,83 @@ def submit_transaction(
         return response
 
     result = cast(Dict[str, Any], response.result)
-    raise XRPLRequestFailureException(result["error"], result["error_message"])
+    raise XRPLRequestFailureException(result)
+
+
+def _prepare_transaction(
+    transaction: Transaction,
+    wallet: Wallet,
+) -> Dict[str, Any]:
+    """
+    Prepares a Transaction by converting it to a JSON-like dictionary, converting the
+    field names to CamelCase. If a Client is provided, then it also autofills any
+    relevant fields.
+
+    Args:
+        transaction: the Transaction to be prepared.
+        wallet: the wallet that will be used for signing.
+
+    Returns:
+        A JSON-like dictionary that is ready to be signed.
+
+    Raises:
+        XRPLException: if both LastLedgerSequence and `ledger_offset` are provided, or
+            if an address tag is provided that does not match the X-Address tag.
+    """
+    transaction_json = transaction_json_to_binary_codec_form(transaction.to_dict())
+    transaction_json["SigningPubKey"] = wallet.pub_key
+
+    _validate_account_xaddress(transaction_json, "Account", "SourceTag")
+    if "Destination" in transaction_json:
+        _validate_account_xaddress(transaction_json, "Destination", "DestinationTag")
+
+    # DepositPreauth
+    _convert_to_classic_address(transaction_json, "Authorize")
+    _convert_to_classic_address(transaction_json, "Unauthorize")
+    # EscrowCancel, EscrowFinish
+    _convert_to_classic_address(transaction_json, "Owner")
+    # SetRegularKey
+    _convert_to_classic_address(transaction_json, "RegularKey")
+
+    return transaction_json
+
+
+def _autofill_transaction(transaction: Transaction, client: Client) -> Transaction:
+    transaction_json = transaction.to_dict()
+    if "sequence" not in transaction_json:
+        sequence = get_next_valid_seq_number(transaction_json["account"], client)
+        transaction_json["sequence"] = sequence
+    if "fee" not in transaction_json:
+        transaction_json["fee"] = get_fee(client)
+
+    if "last_ledger_sequence" not in transaction_json:
+        ledger_sequence = get_latest_validated_ledger_sequence(client)
+        transaction_json["last_ledger_sequence"] = ledger_sequence + _LEDGER_OFFSET
```

Comment:
Future-feature: clients should be able to explicitly specify that they don't want a last ledger sequence on a transaction. ("I want all the defaults except don't let my transaction expire. Also, I want a number 4, no pickles.")